GitCommands/Git/GitGpgController.cs

```diff
@@ -0,0 +1,297 @@
+        string GetCommitVerificationMessage();
+
+        /// <summary>
+        /// Obtain the tag status on current revision.
+        /// </summary>
+        /// <returns>Enum value that indicate if current git revision has one tag with good signature, one tag with bad signature or more than one tag.</returns>
+        GitGpgController.TagStatus CheckTagSign();
+
+        /// <summary>
+        /// Accessor for Revision property.
+        /// </summary>
+        GitRevision Revision { get; set; }
+
+        /// <summary>
+        /// Accessor for Module property.
+        /// </summary>
+        IGitModule Module { get; set; }
+
+        /// <summary>
+        /// Obtain the number of tag on current git revision.
+        /// </summary>
+        /// <returns>Returns the number of annotated or signed tag on current revision.</returns>
+        int NumberOfTag { get; }
+
+        /// <summary>
+        /// Obtain the tag verification message for all the tag in current git revision 
+        /// </summary>
+        /// <returns>Full concatenated string coming from GPG analysis on all tags on current git revision.</returns>
+        string TagVerifyMessage { get; }
+    }
+
+
+    public class GitGpgController : IGitGpgController
+    {
+        private IGitModule _module;
+        private GitRevision _revision;
+
+        /* Commit GPG status */
+        private const string GOOD_SIGN = "G";
+        private const string BAD_SIGN = "B";
+        private const string UNK_SIGN_VALIDITY = "U";
+        private const string EXPIRED_SIGN = "X";
+        private const string EXPIRED_SIGN_KEY = "Y";
+        private const string REVOKED_KEY = "R";
+        private const string MISSING_PUB_KEY = "E";
+        private const string NO_SIGN = "N";
+
+
+        /* Tag GPG status */
+        private const string _goodSignature = "GOODSIG";
+        private const string _validTagSign = "VALIDSIG";
+        private const string _noTagPubKey = "NO_PUBKEY";
+
+
+        /// <summary>
+        /// Obtain the number of tag on current git revision.
+        /// </summary>
+        /// <returns>Returns the number of annotated or signed tag on current revision.</returns>
+        public int NumberOfTag { get; private set; }
+
+        /// <summary>
+        /// Obtain the tag verification message for all the tag in current git revision 
+        /// </summary>
+        /// <returns>Full concatenated string coming from GPG analysis on all tags on current git revision.</returns>
+        public string TagVerifyMessage { get; private set; }
+
+
+        public enum CommitStatus
+        {
+            GoodSignature = 0,
+            SignatureError = 1,
+            MissingPublicKey = 2,
+            NoSignature = 3
+        };
+
+        public enum TagStatus
+        {
+            OneGood = 0,
+            OneBad = 1,
+            Many = 2,
+            NoPubKey = 3
+        };
+
+
+        public GitGpgController()
+        {
+            NumberOfTag = 0;
+        }
```

Comment:
Overall it looks good.
I am somewhat concerned with setting `Module` and `Revision` outside the constructor. This potentially exposes the controller to side effects because the controller is stateful - there is nothing preventing properties from being changed between method calls. 
If the module and the revision are supplied via constructor, the controller is resilient to unexpected change.
It will also make the implementation simpler - no need for `CheckVariable()` method, no need for public properties (`Revision` and `Module`).

The price of immutability - the controller will need to be instantiated for every revision, but I don't think it is a major issue, the controller's memory foot print should be pretty light.

What do you think?

Reply:
The first implementation of this controllew was done in that way, but during execution I've seen that `GitGpgController` constructor's input parameter Module was null, maybe because the caller class `CommitInfo` is a partial class.
This was the only way I've found to set Module and Revision instead of constructor parameter.

Let me know what do you think about.

Reply:
Here's an example how you can supply the parameters via the constructor https://github.com/gpongelli/gitextensions/pull/2

Reply:
Merged with other changes.